Hide Function Messages

AdritRao opened the pull request:
<!--

This source file is part of the Stanford Spezi open-source project

SPDX-FileCopyrightText: 2022 Stanford University and the project authors (see CONTRIBUTORS.md)

SPDX-License-Identifier: MIT

-->

# Hide Function Messages

## :recycle: Current situation & Problem
Only system messages are being hidden. Function messages should also be hidden in the UI. 

## :bulb: Proposed solution
Added support to hide function messages from the view using the same method as hiding system messages.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -19,6 +19,7 @@ public struct MessagesView: View {
     @Binding var chat: [Chat]
     @Binding var bottomPadding: CGFloat
     let hideSystemMessages: Bool
+    let hideFunctionMessages: Bool
```
I noted that we don't forward these properties to the `MessageView`. This is a bug in the current implementation, but we should solve it now as we already touch this system.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -14,6 +14,7 @@ import SwiftUI
 public struct MessageView: View {
     let chat: Chat
     let hideSystemMessages: Bool
+    let hideFunctionMessages: Bool
```
I think it would rather make sense to make the API extensible for the future and even beyond different messages. E.g. using a dynamic set of roles could be a good way to enable the functionality:
```suggestion
    let hideMessagesWithRoles: Set<Chat.Role>
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -68,29 +69,35 @@ public struct MessagesView: View {
         _ chat: Binding<[Chat]>,
         bottomPadding: Binding<CGFloat> = .constant(0),
-        hideSystemMessages: Bool = true
+        hideSystemMessages: Bool = true,
+        hideFunctionMessages: Bool = true
```
I think it would be good to share a constant, that defines what messages should be hidden and is configured in a single place. If we follow the approach to use a `Set`, we can define something like this:
```swift
/// Displays the content of a `Chat` message in a message bubble
public struct MessageView: View {
    public enum Defaults {
        static let hideMessagesWithRoles: Set<Chat.Role> = [.system, .function]
    }
    
    // ...
```

You can then use the `MessageView.Defaults.hideMessagesWithRoles` constant in both the initializers of the `MessagesView` and `MessageView` to provide a good mechanism to define this once and share this across views.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -38,7 +39,10 @@ public struct MessageView: View {
     
     public var body: some View {
-        if chat.role != .system || (chat.role == .system && !hideSystemMessages), let content = chat.content {
+        if chat.role != .system ||
+           (chat.role == .system && !hideSystemMessages) ||
+           (chat.role == .function && !hideFunctionMessages),
+           let content = chat.content {
```
There is an issue with the current implementation: `chat.role != .system` ensures that `.function` messages are always displayed. By using a set, we can greatly simplify this statement, e.g. like this:
```suggestion
       if !hideMessagesWithRoles.contains(chat.role), let content = chat.content {
```

PSchmiedmayer reviewed (changes requested):
Thank you for the PR @AdritRao!

I had some comments and suggestions for improvements in the different files. Be sure that you e.g. add a UI test to the UI test target that adds an `.function` message in `ContentView.swift` and ensures that it is now shown on screen. We can extend `testSpeziMLChat()` in `TestAppUITests.swift` to check that neither the system nor the function call message is shown on screen while the assistant and user messages are displayed.

I think it would also make sense to add a simple new argument to the
```swift
public func queryAPI(withChat chat: [Chat]) async throws -> AsyncThrowingStream<ChatStreamResult, Error>
```
function that would also take a `[ChatFunctionDeclaration]` argument to ensure that we can easily use the current API surface in our LLMonFHIR PRs. That might be a minimal change that we can do here but could make using the `OpenAIComponent` way easier.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageInputView.swift:
```diff
@@ -101,7 +102,7 @@ struct MessageInputView_Previews: PreviewProvider {
                 .ignoresSafeArea()
             VStack {
-                MessagesView($chat)
+                MessagesView($chat, hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
```
if we use the default value in the initializer, we can remove the argument here.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -67,27 +63,27 @@ public struct MessageView: View {
         }
     }
     
+    public enum Defaults {
+        static let hideMessagesWithRoles: Set<Chat.Role> = [.system, .function]
+    }
```
You can address the SwiftLint errors by moving this nested type at the beginning of the enclosing type. You can find out more about the rule here: https://realm.github.io/SwiftLint/type_contents_order.html

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -36,9 +34,8 @@ public struct MessageView: View {
         chat.allignment == .leading ? -7 : 7
     }
     
-    
```
Nice, good catch!

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -67,27 +63,27 @@ public struct MessageView: View {
-    ///   - chat: The chat message that should be displayed.
-    ///   - hideSystemMessages: If system messages should be hidden from the chat overview.
-    public init(_ chat: Chat, hideSystemMessages: Bool = true) {
+    public init(_ chat: Chat, hideMessagesWithRoles: Set<Chat.Role>) {
```
You can use default arguments like this to ensure that you don't have to pass this argument in all the other views: 
```suggestion
    public init(_ chat: Chat, hideMessagesWithRoles: Set<Chat.Role> = MessageView.Defaults.hideMessagesWithRoles) {
```

This also means that these changes are no breaking changes for SpeziML, ensuring that we can tag a new semantic version minor release and don't need to manually update all projects that use SpeziML.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -67,27 +63,27 @@ public struct MessageView: View {
+    public init(_ chat: Chat, hideMessagesWithRoles: Set<Chat.Role>) {
         self.chat = chat
-        self.hideSystemMessages = hideSystemMessages
+        self.hideMessagesWithRoles = MessageView.Defaults.hideMessagesWithRoles
```
This current implementation would ignore everything that is passed in the initializer and just override with the default values:
```suggestion
        self.hideMessagesWithRoles = hideMessagesWithRoles
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -67,27 +63,27 @@ public struct MessageView: View {
     }
 }
 
-
```
Please re-add the two empty lines between the different types in this view.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -67,27 +63,27 @@ public struct MessageView: View {
-                MessageView(Chat(role: .system, content: "System Message (hidden)!"))
-                MessageView(Chat(role: .user, content: "User Message!"))
-                MessageView(Chat(role: .assistant, content: "Assistant Message!"))
+                MessageView(Chat(role: .system, content: "System Message!"), hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
```
I can not see that the current functionality is working if you look at the previews. If you want this to be shown in the previews, you will have to e.g. use something like this and the system message will not be hidden:
```suggestion
                MessageView(Chat(role: .system, content: "System Message!"), hideMessagesWithRoles: [.function])
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -67,27 +63,27 @@ public struct MessageView: View {
-                MessageView(Chat(role: .user, content: "User Message!"))
-                MessageView(Chat(role: .assistant, content: "Assistant Message!"))
+                MessageView(Chat(role: .system, content: "System Message!"), hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
+                MessageView(Chat(role: .system, content: "System Message (hidden)!"), hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
+                MessageView(Chat(role: .function, content: "Function Message!"), hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
+                MessageView(Chat(role: .user, content: "User Message!"), hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
+                MessageView(Chat(role: .assistant, content: "User Message!"), hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
```
After using the default value in the initializer parameter, you can probably use the explicit arguments here.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -9,12 +9,10 @@
     let chat: Chat
-    let hideSystemMessages: Bool
-    
+    let hideMessagesWithRoles: Set<Chat.Role>
```
Please add two empty lines between stored properties and computed properties:
```suggestion
    let hideMessagesWithRoles: Set<Chat.Role>
    
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -9,12 +9,10 @@
 import OpenAI
 import SwiftUI
 
-
```
Please re-add the two empty lines between the import statements and the main type.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -43,7 +42,7 @@ public struct MessagesView: View {
                 VStack {
                     ForEach(Array(chat.enumerated()), id: \.offset) { _, message in
-                        MessageView(message)
+                        MessageView(message, hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles)
```
If you want to enable the functionality to pass the argument down to the message views, you will need to pass the arguments down to the individual view, not the default values:
```suggestion
                        MessageView(message, hideMessagesWithRoles: hideMessagesWithRoles)
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
     ///   - chat: The chat messages that should be displayed.
-    ///   - bottomPadding: A bottom padding for the messages view.
-    ///   - hideSystemMessages: If system messages should be hidden from the chat overview.
+    ///   - bottomPadding: A fixed bottom padding for the messages view.
```
If you can pass the padding is a binding, it is not fixed but variable:
```suggestion
    ///   - bottomPadding: A bottom padding for the messages view.
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
-    ///   - bottomPadding: A bottom padding for the messages view.
-    ///   - hideSystemMessages: If system messages should be hidden from the chat overview.
+    ///   - bottomPadding: A fixed bottom padding for the messages view.
+    ///   - hideMessagesWithRoles: If .system and/or .function messages should be hidden from view.
```
This description is not fully accurate, as this is defined by the argument that is passed in here. I would rather keep it more general, e.g. something like: Defines which messages should be hidden based on the passed in message roles.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
         _ chat: Binding<[Chat]>,
-        bottomPadding: Binding<CGFloat> = .constant(0),
-        hideSystemMessages: Bool = true
+        hideMessagesWithRoles: Set<Chat.Role>,
```
I would also suggest to add a default value here, similar to the comment in the `Message` view.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
         _ chat: [Chat],
-        bottomPadding: CGFloat = 0,
-        hideSystemMessages: Bool = true
+        hideMessagesWithRoles: Set<Chat.Role>,
```
I would also suggest to add a default value here, similar to the comment in the `Message` view.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
     ///   - chat: The chat messages that should be displayed.
     ///   - bottomPadding: A fixed bottom padding for the messages view.
-    ///   - hideSystemMessages: If system messages should be hidden from the chat overview.
+    ///   - hideMessagesWithRoles: If .system and/or .function messages should be hidden from view.
```
Similar feedback as in the initializer below.

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -15,10 +15,9 @@ import SwiftUI
     
-    
     @Binding var chat: [Chat]
     @Binding var bottomPadding: CGFloat
-    let hideSystemMessages: Bool
+    let hideMessagesWithRoles: Set<Chat.Role>
```
We can make all these properties `private`.

PSchmiedmayer commented on Sources/SpeziOpenAI/ChatView.swift:
```diff
@@ -23,7 +23,7 @@ public struct ChatView: View {
         ZStack {
             VStack {
-                MessagesView($chat, bottomPadding: $messageInputHeight)
+                MessagesView($chat, hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles, bottomPadding: $messageInputHeight)
```
If we use a default value in the initializers, we can remove the argument here:
```suggestion
                MessagesView($chat, bottomPadding: $messageInputHeight)
```

PSchmiedmayer commented on Sources/SpeziOpenAI/OpenAIComponent.swift:
```diff
@@ -71,13 +71,13 @@ public class OpenAIComponent: Component, ObservableObject, ObservableObjectProvi
     ///
     /// - Returns: The content of the response from the API.
-    public func queryAPI(withChat chat: [Chat]) async throws -> AsyncThrowingStream<ChatStreamResult, Error> {
+    public func queryAPI(withChat chat: [Chat], withFunction chatFunctionDeclaration: [ChatFunctionDeclaration]) async throws -> AsyncThrowingStream<ChatStreamResult, Error> {
```
I think it might make sense to add a default value here as well, so this is not a breaking API change. We can just use a default empty Array:
```suggestion
    public func queryAPI(withChat chat: [Chat], withFunction chatFunctionDeclaration: [ChatFunctionDeclaration] = []) async throws -> AsyncThrowingStream<ChatStreamResult, Error> {
```

PSchmiedmayer commented on Tests/UITests/TestAppUITests/TestAppUITests.swift:
```diff
@@ -77,6 +77,9 @@ class TestAppUITests: XCTestCase {
         XCTAssert(app.staticTexts["Assistant Message!"].waitForExistence(timeout: 2))
         XCTAssert(app.buttons["Arrow Up Circle"].waitForExistence(timeout: 2))
         
+        XCTAssertFalse(app.staticTexts["System Message!"].waitForExistence(timeout: 2))
+        XCTAssertFalse(app.staticTexts["Function Message!"].waitForExistence(timeout: 2))
+        
```
Thank you for improving the tests! 🚀

PSchmiedmayer reviewed (changes requested):
Thank you for the improvements @AdritRao!

I had a few comments there and there to ensure that we make this PR a non-breaking change for SpeziML and that the functionality of providing arguments is actually passed down to the individual `MessageView`. 

Let me know when you want an other review, thank you for the time and effort that goes into the Swift Package!

AdritRao commented:
Hi @PSchmiedmayer, thank you for your feedback on the PR! I have addressed the comments and fixed the SwiftLint issues. Looking forward to adding this feature to Spezi ML!

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -12,9 +12,12 @@ import SwiftUI
+    public enum Defaults {
+        public static let hideMessagesWithRoles: Set<Chat.Role> = [.system, .function]
+    }
     
```
Let's add two empty lines between the nested type and the stored properties:
```suggestion
    
    
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -12,9 +12,12 @@ import SwiftUI
+        public static let hideMessagesWithRoles: Set<Chat.Role> = [.system, .function]
+    }
     
+    let chat: Chat
```
You can make this property private:
```suggestion
    private let chat: Chat
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -67,13 +68,9 @@ public struct MessageView: View {
     }
     
-    
-    /// - Parameters:
-    ///   - chat: The chat message that should be displayed.
-    ///   - hideSystemMessages: If system messages should be hidden from the chat overview.
```
Please ensure that you leave and have documentation for any `public` method. This makes it easier for someone to dig through the code and the documentation on the Swift Package index to find and understand the method they are looking for.

You can e.g., reuse some wording from `Sources/SpeziOpenAI/MessagesView.swift` (e.g., for the `hideMessagesWithRoles` parameter)

PSchmiedmayer commented on Sources/SpeziOpenAI/MessageView.swift:
```diff
@@ -82,12 +79,13 @@ struct MessageView_Previews: PreviewProvider {
-                MessageView(Chat(role: .system, content: "System Message (hidden)!"))
-                MessageView(Chat(role: .user, content: "User Message!"))
-                MessageView(Chat(role: .assistant, content: "Assistant Message!"))
+                MessageView(Chat(role: .system, content: "System Message!"), hideMessagesWithRoles: [.system])
+                MessageView(Chat(role: .system, content: "System Message (hidden)!"), hideMessagesWithRoles: [.system])
+                MessageView(Chat(role: .function, content: "Function Message!"), hideMessagesWithRoles: [.function])
+                MessageView(Chat(role: .user, content: "User Message!"), hideMessagesWithRoles: [.user])
+                MessageView(Chat(role: .assistant, content: "User Message!"), hideMessagesWithRoles: [.user])
```
If you check out the preview, it seems like all of the messages are hidden (except the .assistant message).
Something like this might make more sense and allows you to visually test the behaviour:
```suggestion
                MessageView(Chat(role: .system, content: "System Message!"), hideMessagesWithRoles: [])
                MessageView(Chat(role: .system, content: "System Message (hidden)!"))
                MessageView(Chat(role: .function, content: "Function Message!"), hideMessagesWithRoles: [.system])
                MessageView(Chat(role: .user, content: "User Message!"))
                MessageView(Chat(role: .assistant, content: "Assistant Message!"))
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
         _ chat: [Chat],
-        bottomPadding: CGFloat = 0,
-        hideSystemMessages: Bool = true
+        hideMessagesWithRoles: Set<Chat.Role> = [.system, .function],
```
As discussed in the previous review, you can also use the default parameter here:
```suggestion
        hideMessagesWithRoles: Set<Chat.Role> = MessageView.Defaults.hideMessagesWithRoles,
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
     }
-    
-    
+
```
Thank you!

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -67,32 +66,31 @@ public struct MessagesView: View {
         _ chat: Binding<[Chat]>,
-        bottomPadding: Binding<CGFloat> = .constant(0),
-        hideSystemMessages: Bool = true
+        hideMessagesWithRoles: Set<Chat.Role> = [.system, .function],
```
As discussed in the previous review, you can also use the default parameter here:
```suggestion
        hideMessagesWithRoles: Set<Chat.Role> = MessageView.Defaults.hideMessagesWithRoles,
```

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -108,9 +106,10 @@ struct MessagesView_Previews: PreviewProvider {
                 Chat(role: .user, content: "User Message!"),
                 Chat(role: .assistant, content: "Assistant Message!")
-            ]
+            ], hideMessagesWithRoles: MessageView.Defaults.hideMessagesWithRoles
```
If you use the default parameters above, you could omit the argument here:
```suggestion
            ]
```

PSchmiedmayer commented on Sources/SpeziOpenAI/OpenAIComponent.swift:
```diff
@@ -71,13 +71,16 @@ public class OpenAIComponent: Component, ObservableObject, ObservableObjectProvi
     ///
     /// - Returns: The content of the response from the API.
-    public func queryAPI(withChat chat: [Chat]) async throws -> AsyncThrowingStream<ChatStreamResult, Error> {
+    public func queryAPI(
+        withChat chat: [Chat],
+        withFunction chatFunctionDeclaration: [ChatFunctionDeclaration] = []
+    ) async throws -> AsyncThrowingStream<ChatStreamResult, Error> {
```
Nice, thank you for the changes! 🚀

PSchmiedmayer commented on Sources/SpeziOpenAI/MessagesView.swift:
```diff
@@ -15,10 +15,9 @@ import SwiftUI
     private static let bottomSpacerIdentifier = "Bottom Spacer"
     
-    
     @Binding var chat: [Chat]
     @Binding var bottomPadding: CGFloat
```
In line with the other property, you can also make these private:
```suggestion
    @Binding private var chat: [Chat]
    @Binding private var bottomPadding: CGFloat
```

PSchmiedmayer reviewed (changes requested):
Thank you for the improvements, we are getting close to merge the PR 🚀

I just had a few comments left.

**One hint**: Any comment with a `suggestion` can be simply accepted and put right in the diff by pressing the button below the GitHub review. Might just make your life a bit easier if you accept suggestions in these comments. 

You can then `git pull` these changes to your local branch 👍

vishnuravi commented:
Great work with this PR @AdritRao!

PSchmiedmayer reviewed (approved):
Thank you @AdritRao, the PR looks great. Ready to be merged, thank you for all the time and effort that went into the PR 🚀
